process: delay and simplify the setup of async hooks trace events

joyeecheung opened the pull request:
#### src: move async hooks trace events setup to pre_execution.js

Reasons:

- Moves more environment-dependent setup out of bootstrap/node.js
- No async operations should be done before the call to
  the setup functions in pre_execution.js so no async hooks should be
  triggered before that. Therefore it is safe to delay the setup
  until then.

#### process: simplify the setup of async hooks trace events

- Remove `trace_category_state` from `Environment` - since this is
  only accessed in the bootstrap process and later in the
  trace category update handler, we could just pass the initial
  values into JS land via the trace_events binding, and pass
  the dynamic values directly to the handler later, instead of
  accessing them out-of-band via the AliasedBuffer.
- Instead of creating the hooks directly in
  `trace_events_async_hooks.js`, export the hook factory and
  create the hooks in trace category state toggle.

#### test: add test for node.async_hooks tracing in workers

#### test: add test for dynamically enabling node.async_hooks tracing

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/20747/

jasnell commented:
> Trace events are not available in worker threads.

I think I missed something. Why is this? They should be available.

addaleax commented:
@jasnell Right – it’s just the JS module that’s not accessible… Is there a test for this code? I think that should have failed in CI if this PR breaks trace_events + Workers, but it looks like everything passed here

joyeecheung commented:
@addaleax I don't think there is a test, it could be added but I don't think the dynamic switch can be tested in workers since dynamically enabling the category requires the trace_events module. I could add the preparation code to the worker bootstrap script though, at least that allows us to test the initial enabling of the events.

joyeecheung commented:
I added the setup to the worker bootstrap, fixed a bug in the C++ part (when the category is enabled the value is 0x7, not 1, so we should test for `!= 0` instead), and added two tests. @jasnell @addaleax PTAL.

CI: https://ci.nodejs.org/job/node-test-pull-request/20879/

joyeecheung commented:
Added hasTracing guards in tests. Not sure why the CI complained about missing dynamically linked deps and libc, let's see what happens in a new CI: https://ci.nodejs.org/job/node-test-pull-request/20907/

richardlau commented on test/parallel/test-trace-events-async-hooks-dynamic.js:
```diff
@@ -0,0 +1,83 @@
+    env: {
+      'NODE_DEBUG_NATIVE': 'tracing',
+      'NODE_DEBUG': 'tracing'
+    }
```
This will fail on some platforms (e.g. Linuxone) in the CI where `LD_LIBRARY_PATH` is used to point to the C++ runtime libraries. The tests should extend `env` rather than replace it.

richardlau commented on test/parallel/test-trace-events-async-hooks-worker.js:
```diff
@@ -0,0 +1,92 @@
+  { cwd: tmpdir.path, env: {
+    'NODE_DEBUG_NATIVE': 'tracing',
+    'NODE_DEBUG': 'tracing'
+  }
```
This will fail on some platforms (e.g. Linuxone) in the CI where `LD_LIBRARY_PATH` is used to point to the C++ runtime libraries. The tests should extend `env` rather than replace it.

joyeecheung commented:
@richardlau Thanks! I've updated the `env` passed to `spawnSync`. CI: https://ci.nodejs.org/job/node-test-pull-request/20913/

joyeecheung commented:
Landed in dec3dad...85df2c4
